### DISCUSSION TITLE
Designer - order of execution for multiple objects.

### DISCUSSION THREAD

**Discussion by dmilton** [winder/Universal-G-Code-Sender#2189] [Help] [answered]
Hello,

I'm new to UGS Platform and CNC in general so this may be obvious to some but I'm having difficulty getting design elements to run/execute in the order I would like them to.

I have a project which has many (31) design elements. All of the design elements are ordered with the expectation that cutting/milling will take place in the order the design elements are listed. I've re-created many of these elements in a separate files and tested on a scrap piece of wood and they behave as expected but these are essentially 2 or 3 element tests.

The problem is when I run the full project the elements get run almost completely out of order.  Down to the point where I cut two pockets - one is 2.4mm deep and 36mm square and the second one is inset 2.5mm from two edges of the first and is 27.7mm by 23.9mm and is 5mm deep. When I send this to my 3040 prover CNC the first run had an element order of execution of 3, 4, 1, 2, 29, 30, 27, 28, 15, 16, 14, 13, 18, 17, 31 at which point I aborted because 31 is the last step and cuts the finished part out from the main piece of wood. In a couple of the above instances it cuts air because the the pocket is defined as two parts, a shallower one and a deeper smaller one inside. Unfortunately, the deeper part of the pocket is cut first, then the shallower part is cut afterwards. About 80% of the second shallower pocket cutting is just cutting air and the first pocket makes the CNC rather angry, causing skipping or in one instance cutting so deep it just burns a path.

I've tried exporting the full project to gcode and it seems the same silliness regarding order is being carried on in the exported gcode as well. What I find odd is that looking at the ugsd file all of the design elements are listed in exactly the order I would like them to execute in. Why is it re-ordered on export and for run/execution???

Is this a bug in UGS Platform? Is this a setting I'm missing? Is there something else going on? How do I FORCE an order of execution? I'm currently looking at exporting to gcode and manually re-ordering the elements but I can't help but think this is the kind of thing the UGS was designed to do. Am I missing something really fundamental?

Thanks in advance for you help and suggestions.

**Comment by breiler** [comment, marked as answer]
Hi David,

When I first built the design feature there was so much travel time between different objects so I thought that it would be a good idea to optimise it with some ordering so that it works on the closest objects first. This turned out to be a bad idea for just the reason you have described.

Therefore I changed behaviour in the latest release where it will process the objects in the order defined by the list.

Please try the latest version from here:
https://github.com/winder/Universal-G-Code-Sender/releases/tag/v2.0.15

Thanks for the feedback by the way, the designer tool is quite new so if there is something else that grinds your gears please let me know.

**Reply by dmilton** [reply]
Thanks for the quick fix. Asking for suggestions is a dangerous (although often productive) thing to do :-)

I had not considered proximity as one of the ordering factors but once you say it, I understand why that was done. One way to allow some optimization based on proximity is to allow grouping of elements. Order within the element is considered fixed but between groups the order can be optimized.

I have a pocket that is actually composed of 4 different parts. Two of them need to proceed in order but for the other two order is irrelevant. My project calls for 4 identical pockets, each rotated 90 degrees around the centre. I had originally started writing gcode where the pocket was defined as a subroutine and then I would rotate the coordinate plane (M68/M69) and reposition for each pocket.  Unfortunately, my little GRBL CNC is not able to handle such code so I ended up creating a whole bunch of elements. Having a way for the designer tool to do this would be fantastic! Perhaps by using an element group and allowing the group coordinates set, including rotation.

Another thing I've seen hints of but cannot seem to find is a way to "start here." It would be nice to select an element and have just that single element run. Or have it start from that element and run to the end. That's where I'm at now. I have a partially completed scale that needs two more of my grouped pockets and then the outer cut to be made. It would be nice to select the element to start with and just say "go" and have the rest of the work completed.

Last thing that would be nice; my scale is a circle that will be cut out of a rectangular piece of wood. It would be nice to be able to define "no-cut" points in that circle so perhaps every 90 degrees a couple of mm of material is left behind. What I'm thinking is a small "cube" of material towards the bottom of the circular cut that is skipped. This would leave the the circle at least partially attached to the rectangle of wood. Once finished I could the use a coping saw to cut the last little bits out. I've been trying to figure out how to get the circle cut most of the way through and then define arcs which do the rest but either the capability does not yet exist or I haven't figured out how.

Anyway, thanks for the quick fix on my ordering problem.

Cheers,
  Dave.

**Reply by dmilton** [reply]
I've got a weird problem with the new version. It's pretty clear by looking at the two screen shots. I'm running MacOS 12.6.3 if that makes a difference. Is this going to do the cutting I expect it to?

<img width="382" alt="UGS 2 0 15" src="https://user-images.githubusercontent.com/3662577/229297756-d1a2ed22-d7e0-486e-860b-7eaac5c55ef4.png">
<img width="567" alt="UGS 2 0 14" src="https://user-images.githubusercontent.com/3662577/229297949-6c799c72-736d-450a-b7cd-d650a9c5deeb.png">

**Reply by breiler** [reply]
The images you posted is probably a result for a fix I made to be able to import SVG:s. So I have unfortunately destroyed your design file. It may be salvageable by increasing the size of each rectangle so they get ten times larger.

Regarding the suggestion to be able to make copies that are rotated sounds like a good idea. 

There is no way to "run from" or "run selected" today, and it will be a bit tricky to add. As a workaround you could place the objects in groups and hide/show them.

The last feature to add tabs is on the todo, I have just not gotten around to it yet.

**Reply by dmilton** [reply]
What I found odd is the design file itself is "unchanged" and I can open it in 2.0.14 and it looks fine. In some respects, it's probably just as easy to re-create the entire design in 2.0.15. I just wondered what happened. I'll definitely take the SVG import. My original design is an SVG so it may be easier to import that and adjust the depth as appropriate.
